### PR TITLE
Refactor common code into client.py

### PR DESCRIPTION
**Pull request by bloodearnest**
This is in anticipation of adding repoupdater into this repository

Also refactored how changes where logged/executed in the process, as it was
getting slower and slower due to increasing repository count.

### CODE REVIEW

File: manage-github.py [hunk 1]
```diff
@@ -158,74 +118,95 @@ def ensure_teams(config, org):
...
+    if changes:
+        print("Do you want to apply the above changes (y/n)?")
+        i, _, _ = select.select([sys.stdin], [], [], 5)
+        if i:
+            if sys.stdin.readline().strip().lower() == 'y':
+                for change in changes:
+                    print(change)
+                    change()
+    else:
+        print('No changes needed')
```
Comment (madwort) [inline manage-github.py:206]: not quite sure about this block - if `args.execute` is true, `changes` will be `[]` (falsey), so the path through the code will do
```
print(change)
change()
print('No changes needed')
```

File: manage-github.py [hunk 2]
```diff
@@ -124,11 +81,14 @@ def protect_branch(repo, branch=None, **kwargs):
...
+        if protection: 
+            yield client.Change(
+                lambda: protected_branch.edit_protection(**protection),
+                'setting branch protection on {} to:\n{}',
+                protected_branch.name,
+                ', '.join('{}={}'.format(k, v) for k, v in protection.items()),
+            )
 
 
 def ensure_teams(config, org):
```
Comment (madwort) [inline manage-github.py:94]: two args

File: manage-github.py [hunk 3]
```diff
@@ -158,74 +118,95 @@ def ensure_teams(config, org):
...
+    org = client.get_org()
+    config = yaml.safe_load(open(args.config))
 
-    if not EXECUTE:
+    if args.dry_run:
         print('*** DRY RUN - no changes will be made ***')
 
-    ensure_teams(config, org)
+    changes = []
+    for change in ensure_teams(config, org, args.execute):
```
Comment (madwort) [inline manage-github.py:190]: three args

Comment (madwort) [reply manage-github.py:190]: doesn't run for me - `TypeError: ensure_teams() takes 2 positional arguments but 3 were given` - does it work for you?

Comment (bloodearnest) [reply manage-github.py]: Yeah, I removed the arg as a last minute fix, didn't rerun, my bad

File: manage-github.py [hunk 4]
```diff
@@ -158,74 +118,95 @@ def ensure_teams(config, org):
...
+    if changes:
+        print("Do you want to apply the above changes (y/n)?")
+        i, _, _ = select.select([sys.stdin], [], [], 5)
+        if i:
+            if sys.stdin.readline().strip().lower() == 'y':
+                for change in changes:
+                    print(change)
+                    change()
+    else:
+        print('No changes needed')
```
Comment (madwort) [inline manage-github.py:206]: I think this line is going to be run (incorrectly) if `args.execute` is true

File: manage-github.py (hunk 1 above)
Comment (madwort) [reply manage-github.py]: moving to updated commit

File: .github/workflows/tests.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        python-version: 3.8
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt
+    - name: Run dry-run
+      env:
+        ORG_TOKEN: ${{ secrets.ORG_READONLY_TOKEN }}
+      run: |
+	echo "WARNING: ignore output from this command - it is running w/o elevated priveleges needed to correctly read branch protection details"
```
Comment (madwort) [inline .github/workflows/tests.yaml:25]: typo `priveleges`

File: manage-github.py [hunk 6]
```diff
@@ -158,74 +119,117 @@ def ensure_teams(config, org):
...
+    parser.add_argument('config', help='The team config')
+    parser.add_argument('--exec', action='store_true',
+                        dest='execute',
+                        help='Automatically execute commands')
+    parser.add_argument('--dry-run', action='store_true',
+                        dest='dry_run',
+                        help='Just print what would change and exit')
+
+    args = parser.parse_args(argv)
+    # we run in one of three modes:
```
Comment (madwort) [inline manage-github.py:192]: yeah, this is much clearer to me now

File: .github/workflows/tests.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        python-version: 3.8
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt
+    - name: Run dry-run
+      env:
+        ORG_TOKEN: ${{ secrets.ORG_READONLY_TOKEN }}
+      run: |
+	echo "WARNING: ignore output from this command - it is running w/o elevated privileges needed to correctly read branch protection details"
```
Comment (madwort) [inline .github/workflows/tests.yaml:25]: doh! and YAML syntax typo!

File: manage-github.py (hunk 6 above)
Comment (bloodearnest) [reply manage-github.py:192]: right, I realised there were some corner cases if you specified both --dry-run --exec - this makes so there's only ever one mode in operation, and --dry-run>>--exec

File: .github/workflows/tests.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt
+    - name: Run dry-run
+      env:
+        ORG_TOKEN: ${{ secrets.ORG_READONLY_TOKEN }}
+      run: |
+        echo "WARNING: ignore output from this command - it is running w/o elevated privileges needed to correctly read branch protection details"
+	python manage-github.py config.yaml --dry-run
```
Comment (madwort) [inline .github/workflows/tests.yaml:26]: one more - YAML indentation!

**Review by madwort — APPROVED** [review]
🎉 generators FTW
